KEYCLOAK-849-session-limiting-feature

mfdewit opened the pull request:
Under contract of Royal Boom Publishers I've implemented two authenticators which can be used to limit the amount of sessions for a realm, for a client or for users per client.
Here is a link to the ticket: https://issues.redhat.com/browse/KEYCLOAK-849

mfdewit committed:
All required code for session limiting and integration testing

mposolda commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/sessionlimits/UserSessionLimitsTest.java:
```diff
@@ -0,0 +1,145 @@
+import org.keycloak.testsuite.pages.AppPage;
+import org.keycloak.testsuite.pages.ErrorPage;
+
+public class UserSessionLimitsTest extends AbstractTestRealmKeycloakTest {
```
I believe this will need to have something like `@AuthServerContainerExclude(REMOTE)` at the top as AFAIK the setup of authentication flows with `testingClient.server.run()` probably won't work on the remote server.

mposolda commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/sessionlimits/UserSessionLimitsTest.java:
```diff
@@ -0,0 +1,145 @@
+        testingClient.server().run(session -> {
+            RealmModel realm = session.realms().getRealmByName("test");
+
+            if (realm.getBrowserFlow().getAlias().equals("parent-flow")) {
```
Minor: This can be simplified a bit with `FlowUtil` class. For example see `BrowserFlowTest` class for the inspiration.

mposolda reviewed (changes requested):
@mfdewit Thanks for the contribution! I am very sorry for the late review. Are you still interested in the contribution? Will it work for you to address review comments I've added?

mposolda commented:
@mfdewit Are you able to follow-up on the mentioned issues anytime soon?

Keycloak team is interested a lot in this feature, so we consider to take your PR and continue with the remaining work. Please reply here if you are already working on the review comments or if you would like to start soon on them. Otherwise we can take-over and continue.

mfdewit commented:
@mposolda Hi Marek, good to see that you started the review! Give me a few days to pick up and continue on the remaining work. I'll plan to work on this later this week and keep you posted.

mposolda commented:
@mfdewit Thanks! I will re-review once you add the changes

mposolda commented on services/src/main/java/org/keycloak/authentication/authenticators/sessionlimits/UserSessionLimitsAuthenticator.java:
```diff
@@ -0,0 +1,102 @@
+            ClientModel currentClient = context.getAuthenticationSession().getClient();
+            logger.debugf("session-limiter's current keycloak clientId: %s", currentClient.getClientId());
+
+            List<UserSessionModel> userSessionsForClient = userSessionsForRealm.stream().filter(session -> session.getAuthenticatedClientSessionByClient(currentClient.getId()) != null).collect(Collectors.toList());
```
Additional point: When the count of clients per user ( USER_CLIENT_LIMIT ) is disabled (which IMO should be default behaviour as IMO most common use-case is limit of total user sessions per user), it will be nice to make the code in a way that call to `session.getAuthenticatedClientSessionByClient` is avoided as this has some additional performance penalty.

mfdewit replied:
Yes, good point! It is now called conditionally.

mposolda commented on services/src/main/java/org/keycloak/authentication/authenticators/sessionlimits/UserSessionLimitsAuthenticatorFactory.java:
```diff
@@ -0,0 +1,109 @@
+    public List<ProviderConfigProperty> getConfigProperties() {
+        ProviderConfigProperty userRealmLimit = new ProviderConfigProperty();
+        userRealmLimit.setName(USER_REALM_LIMIT);
+        userRealmLimit.setLabel("Maximum concurrent sessions for each user");
```
One additional point: IMO it will be nice if USER_CLIENT_LIMIT is disabled by default (hence having 0 or negative value). Not sure what is the best default value for USER_REALM_LIMIT, but maybe it will be nice to add some default value. How about 3? I am leaving to you what you consider to be best default...

mfdewit replied:
Done as requested; the tooltips now mention how to disable the limit. 
USER_CLIENT_LIMIT is disabled by default and USER_REALM_LIMIT is set to 3. This seems a sensible default to me.

mposolda commented on services/src/main/java/org/keycloak/authentication/authenticators/sessionlimits/UserSessionLimitsAuthenticatorFactory.java:
```diff
@@ -0,0 +1,109 @@
+
+        ProviderConfigProperty userClientLimit = new ProviderConfigProperty();
+        userClientLimit.setName(USER_CLIENT_LIMIT);
+        userClientLimit.setLabel("Maximum concurrent sessions for each user per keycloak client");
```
I wonder that  it can be nice if tooltip label has some mention that this value should be always lower than USER_REALM_LIMIT (Unless USER_REALM_LIMIT is disabled) WDYT?

mposolda reviewed (commented):
@mfdewit I've added few more minor points related to the tooltip, default values for authenticator option and possible performance optimization. WDYT?

mfdewit committed:
Processing review comments

mfdewit committed:
Replaced the session limit exceeded error with a generic error

mfdewit commented on services/src/main/java/org/keycloak/authentication/AuthenticationProcessor.java:
```diff
@@ -788,6 +788,11 @@ public Response handleBrowserException(Exception failure) {
                 event.error(Errors.INVALID_USER_CREDENTIALS);
                 if (e.getResponse() != null) return e.getResponse();
                 return ErrorPage.error(session, authenticationSession, Response.Status.BAD_REQUEST, Messages.CREDENTIAL_SETUP_REQUIRED);
+            } else if (e.getError() == AuthenticationFlowError.SESSION_LIMIT_EXCEEDED) {
```
Yes, I agree a generic error would be beneficial in the long run. 
I think I've implemented it the way you intended. Could you please verify?

mfdewit commented on services/src/main/java/org/keycloak/authentication/authenticators/sessionlimits/RealmSessionLimitsAuthenticator.java:
```diff
@@ -0,0 +1,52 @@
+import org.keycloak.events.Errors;
+import org.keycloak.services.messages.Messages;
+
+public class RealmSessionLimitsAuthenticator extends AbstractSessionLimitsAuthenticator {
```
I've double checked with my client who requested the limiting feature and they are not using the Realm limitation either. So it seems you were correct questioning its usefulness. It has been removed and the abstract class is merged with the UserSessionLimitsAuthenticator.

mfdewit commented on services/src/main/java/org/keycloak/authentication/authenticators/sessionlimits/UserSessionLimitsAuthenticator.java:
```diff
@@ -0,0 +1,102 @@
+
+    private void logoutOldestSession(List<UserSessionModel> userSessions) {
+        logger.info("Logging out oldest session");
+        Optional<UserSessionModel> oldest = userSessions.stream().sorted(Comparator.comparingInt(UserSessionModel::getStarted)).findFirst();
```
Fully agree here!

mfdewit committed:
Replaced the session limit exceeded error with a generic error

mfdewit committed:
Set sensible presets

mposolda commented on server-spi-private/src/main/java/org/keycloak/authentication/AuthenticationFlowException.java:
```diff
@@ -76,4 +83,12 @@ public Response getResponse() {
     }
+
+    public String getUserErrorMessage() {
+        return userErrorMessage;
```
the `userErrorMessage` is returned here, but never initialized. So this will always return null unless I am missing something? It seems something will also need to be added to the constructor IMO?

mposolda commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/sessionlimits/RealmSessionLimitsTest.java:
```diff
@@ -0,0 +1,121 @@
+import org.keycloak.testsuite.pages.AppPage;
+
+@AuthServerContainerExclude(AuthServerContainerExclude.AuthServer.REMOTE)
+public class RealmSessionLimitsTest extends AbstractFlowTest {
```
I think this test will need to be removed as well together with RealmSessionLimitsAuthenticatorFactory? I guess it doesn't even compile due the RealmSessionLimitsAuthenticator and factory were removed,

mposolda commented on server-spi-private/src/main/java/org/keycloak/authentication/AbstractAuthenticationFlowContext.java:
```diff
@@ -142,6 +142,15 @@
+     * @param response Response that will be sent back to HTTP client
+     * @param eventDetails Details about the error event
+     */
+    void failure(AuthenticationFlowError error, Response response, String eventDetails);
```
This looks good, but did not you forget to push the change also in the `AuthenticationFlowContext` interface? This PR does not contain it.

mposolda reviewed (changes requested):
@mfdewit Thanks for the changes!

I've added few minor comments inline. I believe some of them are maybe just due the fat that you forgot to push some changes to this PR, which you may already have locally on your laptop? But that is just my guess :-)

mposolda commented on services/src/main/java/org/keycloak/authentication/authenticators/sessionlimits/UserSessionLimitsAuthenticatorFactory.java:
```diff
@@ -0,0 +1,111 @@
+
+        ProviderConfigProperty userClientLimit = new ProviderConfigProperty();
+        userClientLimit.setName(USER_CLIENT_LIMIT);
+        userClientLimit.setLabel("Maximum concurrent sessions for each user per keycloak client. Provide a zero or negative value to disable this limit.");
```
WDYT about adding one more comment like: This should be always smaller than User Realm Limit.

Or something like that just to clarify it a bit more?

mfdewit committed:
Provided eventDetails and some other minor improvements

mfdewit committed:
Provided eventDetails and some other minor improvements

mfdewit committed:
fixed an import

mposolda reviewed (changes requested):
@mfdewit Thanks for the additional updates! The PR itself looks good to me, but there are still some failing tests.
- The `UserSessionLimitsTest` is the new test, which fails in GH actions. Hopefully you can reproduce and fix this one in your environment.
- The `ProvidersTest` fails as it needs to be updated always when new authenticator provider is added to the codebase. This is easy update, which just requires to adjust number and add the expected providerId and help text of new provider. See previous updates of this class in GH history for more details.

mfdewit commented:
@mposolda I'm looking to the failing tests. Had some trouble to get them going again on my local system, but now I can reproduce the error. Stay tuned.

mfdewit committed:
fixed the integration test. And some improvements on the tooltips

mfdewit commented:
> @mfdewit Thanks for the additional updates! The PR itself looks good to me, but there are still some failing tests.
> 
>     * The `UserSessionLimitsTest` is the new test, which fails in GH actions. Hopefully you can reproduce and fix this one in your environment.
> 
>     * The `ProvidersTest` fails as it needs to be updated always when new authenticator provider is added to the codebase. This is easy update, which just requires to adjust number and add the expected providerId and help text of new provider. See previous updates of this class in GH history for more details.

I've fixed the UserSessionLimitsTest (well actually I fixed the UserSessionLimitsAuthenticator; I was re-using a stream multiple times) and extended the ProvidersTest to incorporate this new provider. I also updated the  UserSessionLimitsFactory to actually display the tooltips instead of just having al the info in the field labels.

Anyway, I think this should do it!

mposolda commented:
@mfdewit Thanks again!
